### DISCUSSION TITLE
PHP 8 union type in mount method leads to "Call to undefined method ReflectionUnionType::isBuiltin()"

### DISCUSSION THREAD

**Discussion by smares** [livewire/livewire#3403] [7. Bugs (archived)]
### Discussed in https://github.com/livewire/livewire/discussions/3345

<div type='discussions-op-text'>

<sup>Originally posted by **smares** July 16, 2021</sup>
The following code:

```
public function mount(array|object|null $categories = null)
{
    ...
}
```

Produces the following error:

> Livewire\ImplicitlyBoundMethod::getParameterClassName
> vendor/livewire/livewire/src/ImplicitlyBoundMethod.php:107
> **Call to undefined method ReflectionUnionType::isBuiltin()**

Running Livewire 2.5.5 with Laravel 8.5.22 / Framework 8.50.0.</div>

**Comment by fabio-ivona** [comment]
The same error occours for function with union type parameters, i.e:

```php
public function test_method(float|string $param){
    //...
}
```

I think a simple type check in order to prune `ReflectionUnionTypes` would solve this issue:

```
// ImplicitlyBoundMethod.php

public static function getParameterClassName($parameter)
    {
        $type = $parameter->getType();

        if($type === null){
            return null;
        }
        
        if(! $type instanceof \ReflectionNamedType){
            return null;
        }
        
        if($type->isBuiltin()){
            return null;
        }
       
        return $type->getName();
    }
```


another solution would be iterating over union types to search a valid class name:

```
// ImplicitlyBoundMethod.php

public static function getParameterClassName($parameter)
    {
        $type = $parameter->getType();

        if($type === null){
            return null;
        }

        if($type instanceof \ReflectionUnionType){
            foreach ($type->getTypes() as $union_type){
                if(!$type->isBuiltin()){
                    return $type->getName();
                }
                return null;
            }
        }else{
            return (! $type->isBuiltin()) ? $type->getName() : null;
        }
    }
```

@calebporzio what do you think? Can I open a PR to fix this?

**Comment by tanthammar** [comment]
@smares, @fabio-ivona 
I have the same problem when passing a plain object to the mount() method.

```php
public function mount(object $foo){}
```

The problem is related to a change in php7 vs php8.

<img width="984" alt="reflectionUnionType" src="https://user-images.githubusercontent.com/21239634/129404463-e9fa0e8f-01c0-4a29-83c1-8f2fa53526e8.png">

https://www.php.net/manual/en/class.reflectiontype.php

As Livewire v2.5.5 supports php v7|v8,` isBuiltin()` can't be used for `ReflectionUnionType`.

```php
Call to undefined method ReflectionUnionType::isBuiltin()
```

**Reply by fabio-ivona** [reply]
I proposed two implementation to solve the issue (see my previous response) but would like to know what @calebporzio thinks before starting to work on it

**Reply by tanthammar** [reply]
Yes you are correct, sorry if I expressed myself badly. My intention was only to add info to the topic.

**Reply by fabio-ivona** [reply]
It is a good way to revive this discussion, it has been inactive for a while, thanks!

**Reply by tanthammar** [reply]
I made a PR, don't think Caleb has the time to read all discussions :)
#3519
